# Cancer-Immunotherapy-Microbiome-Analysis  
**Code repository for the manuscript:** *"Non-Intestinal Microbial Signatures in Stool as Predictors of Cancer Immunotherapy Outcome"*

![](https://github.com/JeniaOle13/cancer-biomarkers/blob/main/data/sample_map.jpg)
*Global distribution of collected samples (n=624) across countries. Circle size represents sample count per region, while color indicates the proportion of patients R - responsive (or NR - non-responsive) to immunotherapy*


[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)

## ðŸ“– Overview  
This repository contains the computational workflows and scripts used to analyze 624 stool metagenomes from 11 cohorts to identify microbial biomarkers of cancer immunotherapy response. Key findings include:  
- **Blautia species** as keystone taxa linked to immunotherapy success via the Wood-Ljungdahl pathway.  
- **Oral/food-derived microbes** (e.g., *Veillonella*, *Klebsiella*) as predictors of non-response.  
- **Machine learning models** (AUC = 0.72) for predicting outcomes using log-ratio transformed operational genomic unit (OGU) abundances.

## ðŸ“‚ Repository Structure
